farming/logics/traits/master_chef/farming.rs

```diff
@@ -0,0 +1,482 @@
+        let lp_tokens = &self.data::<Data>().lp_tokens;
+        for i in 0..lp_tokens.len() {
+            self._update_pool(i as u32)?;
+        }
+        Ok(())
+    }
+
+    fn _update_pool(&mut self, pool_id: u32) -> Result<(), FarmingError> {
+        let mut pool = self
+            .get_pool_infos(pool_id)
+            .ok_or(FarmingError::PoolNotFound1)?;
+        let current_block = Self::env().block_number();
+        if current_block > pool.last_reward_block {
+            let lp_supply = self._get_lp_supply(pool_id)?;
+            if lp_supply > 0 {
+                let additional_acc_arsw_per_share =
+                    self._calculate_additional_acc_arsw_per_share(&pool, current_block, lp_supply)?;
+                pool.acc_arsw_per_share = pool
+                    .acc_arsw_per_share
+                    .checked_add(additional_acc_arsw_per_share)
+                    .ok_or(FarmingError::AddOverflow8)?;
+            }
+            pool.last_reward_block = current_block;
+            self.data::<Data>().pool_info.insert(pool_id, &pool);
+
+            self._emit_log_update_pool_event(
+                pool_id,
+                pool.last_reward_block,
+                lp_supply,
+                pool.acc_arsw_per_share,
+            );
+        }
+        Ok(())
+    }
+
+    fn _calculate_additional_acc_arsw_per_share(
+        &mut self,
+        pool_info: &Pool,
+        current_block: u32,
+        lp_supply: Balance,
+    ) -> Result<Balance, FarmingError> {
+        if lp_supply == 0 {
+            return Err(FarmingError::LpSupplyIsZero)
+        }
+        let last_reward_block_period = self._get_period(pool_info.last_reward_block)?;
+        let current_period = self._get_period(Self::env().block_number())?;
+
+        let mut arsw_reward: Balance = 0;
+        let mut last_block = pool_info.last_reward_block;
+        let mut period = last_reward_block_period;
+        while period <= current_period {
+            if period > MAX_PERIOD {
+                break
+            }
+            let total_alloc_point: u32 = self.data::<Data>().total_alloc_point;
+            if current_block <= self._period_max(period)? {
+                arsw_reward = arsw_reward
+                    .checked_add(
+                        (current_block as u128)
+                            .checked_sub(last_block as u128)
+                            .ok_or(FarmingError::SubUnderflow4)?
+                            .checked_mul(self._arsw_per_block(period)?)
+                            .ok_or(FarmingError::MulOverflow3)?
+                            .checked_mul(pool_info.alloc_point as u128)
+                            .ok_or(FarmingError::MulOverflow4)?
+                            .into(),
+                    )
+                    .ok_or(FarmingError::AddOverflow6)?
+                    .checked_div(total_alloc_point.into())
+                    .ok_or(FarmingError::DivByZero1)?
```

Comment:
```suggestion
                        casted_mul(
                            (current_block
                                .checked_sub(last_block)
                                .ok_or(FarmingError::SubUnderflow4)?
                                as u128)
                                * (pool_info.alloc_point as u128),
                            self._arsw_per_block(period)?,
                        )
                        .checked_div(total_alloc_point.into())
                        .ok_or(FarmingError::DivByZero1)?
                        .try_into()
                        .map_err(|_| FarmingError::CastTou128Error1)?,
                    )
                    .ok_or(FarmingError::AddOverflow6)?;
```
Division should be done to addend, not the sum. Plus, by reordering multiplication smartly, we can yield high precision calculation.